lib/improver/spotdata_new/apply_lapse_rate.py

```diff
@@ -0,0 +1,168 @@
+
+"""Apply temperature lapse rate adjustments to a spot data cube."""
+
+import numpy as np
+
+import iris
+from improver.spotdata_new.spot_extraction import SpotExtraction
+
+
+class SpotLapseRateAdjust:
+    """
+    Adjusts spot data temperatures by a lapse rate to better represent the
+    conditions at their altitude that may not be captured by the model
+    orography.
+    """
+
+    def __init__(self, neighbour_selection_method='nearest',
+                 grid_metadata_identifier='mosg'):
+        """
+        Args:
+            neighbour_selection_method (str):
+                The neighbour cube may contain one or several sets of grid
+                coordinates that match a spot site. These are determined by
+                the neighbour finding method employed. This keyword is used to
+                extract the desired set of coordinates from the neighbour cube.
+            grid_metadata_identifier (str or None):
+                A string to search for in the input cube's attributes that
+                can be used to ensure that the cubes being used are for the
+                same model/grid. This test can be bypassed by setting this to
+                None.
+        """
+        self.neighbour_selection_method = neighbour_selection_method
+        self.grid_metadata_identifier = grid_metadata_identifier
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        return ('<SpotLapseRateAdjust: neighbour_selection_method: {}, '
+                'grid_metadata_identifier: {}>'.format(
+                    self.neighbour_selection_method,
+                    self.grid_metadata_identifier))
+
+    def check_grid_match(self, cubes):
+        """
+        Uses the provided grid_metadata_identifier to extract and compare
+        attributes on the input cubes. The expectation is that all the metadata
+        identified should match for the cubes to be deemed compatible.
+
+        Args:
+            cubes (list of cubes):
+                List of cubes for which the attributes should be tested.
+        Raises:
+            ValueError: Raised if the metadata extracted is not identical on
+                        all cubes.
+        """
+        def _get_attributes(cube):
+            """Build dictionary of attributes that match the
+            self.grid_metadata_identifier."""
+
+            attributes = cube.attributes
+            attributes = {k: v for (k, v) in attributes.items()
+                          if self.grid_metadata_identifier in k}
+            return attributes
+
+        def _compare_attributes(attributes, reference_attributes):
+            """Compare keys and values of attributes with reference_attributes.
+            Raise an exception if they do not match exactly."""
+            try:
+                match = ((attributes.items() & reference_attributes.items()) ==
+                         reference_attributes.items())
+            except TypeError:
```

Comment:
This is a python2/3 thing. Using moxie to do any plotting of spot-data output requires we run notebooks with python 2. Python 2 returns a list rather than a view/iterator with a call to .items, and the "&" logic does not work with lists. It is very annoying to have to include this, but I have modified the doc string to indicate it should be removed when we have no more python 2 dependencies.

Reply:
Yes, this certainly isn't ideal. It would be good to have a ticket planned that will remove this.